File: citizensdk/android/native/src/main/kotlin/org/citizen/sdk/internal/CitizenSdkHostRecord.kt
package org.citizen.sdk.internal

/** Exact opaque record completion consumed only by the JNI host bridge. */
internal class CitizenSdkHostRecord(
    val domain: Int,
    val errorCode: Int,
    val present: Boolean,
    val revision: Long,
    record: ByteArray?,
) {
    private val recordValue = record?.clone()
    fun record(): ByteArray? = recordValue?.clone()

    companion object {
        fun absent(domain: Int) = CitizenSdkHostRecord(domain, 0, false, 0, null)
        fun present(domain: Int, revision: Long, record: ByteArray) =
            CitizenSdkHostRecord(domain, 0, true, revision, record)
        fun failure(domain: Int, errorCode: Int) =
            CitizenSdkHostRecord(domain, errorCode, false, 0, null)
    }
}

internal object CitizenSdkHostDomain {
    const val CHAIN_DATABASE = 1
    const val RUNTIME_CACHE = 2
    const val WALLET_PROFILE = 3
    const val TRANSACTION_HISTORY = 4
    const val ENCRYPTED_SECRET_BLOB = 5
}

File: citizensdk/darwin/Sources/CitizenSDK/CitizenSDKHostRecord.swift
import Foundation

internal enum CitizenSDKHostDomain: UInt32 {
    case chainDatabase = 1
    case runtimeCache = 2
    case walletProfile = 3
    case transactionHistory = 4
    case encryptedSecretBlob = 5
}

/// Exact opaque host record copied synchronously by the Rust host adapter.
internal struct CitizenSDKHostRecord {
    let domain: CitizenSDKHostDomain
    let errorCode: CitizenSDKErrorCode
    let present: Bool
    let revision: UInt64
    let record: Data?

    static func absent(_ domain: CitizenSDKHostDomain) -> CitizenSDKHostRecord {
        CitizenSDKHostRecord(domain: domain, errorCode: .ok, present: false, revision: 0, record: nil)
    }

    static func present(_ domain: CitizenSDKHostDomain, revision: UInt64, record: Data) -> CitizenSDKHostRecord {
        CitizenSDKHostRecord(domain: domain, errorCode: .ok, present: true, revision: revision, record: record)
    }

    static func failure(_ domain: CitizenSDKHostDomain, _ error: CitizenSDKErrorCode) -> CitizenSDKHostRecord {
        CitizenSDKHostRecord(domain: domain, errorCode: error, present: false, revision: 0, record: nil)
    }
}
